middleware/src/index.js

```diff
@@ -1,179 +1,271 @@
-        proxy = document.currentScript.getAttribute('data-proxy');
-        token = document.currentScript.getAttribute('data-token');
-        domain = document.currentScript.getAttribute('data-domain');
-        DATASOURCE = document.currentScript.getAttribute('data-datasource') || DATASOURCE;
-       
-        for (const attr of  document.currentScript.attributes) {
-           if( attr.name.startsWith('tb_')) {
-              globalAttributes[attr.name.slice(3)] = attr.value;
-           }
-        }
-        
+  const STORAGE_KEY = 'session-id'
+  let DATASOURCE = 'analytics_events'
+  const storageMethods = {
+    cookie: 'cookie',
+    localStorage: 'localStorage',
+    sessionStorage: 'sessionStorage',
+  }
+  let STORAGE_METHOD = storageMethods.cookie
+  let globalAttributes = {}
+
+  let proxy, token, host, domain
+  if (document.currentScript) {
+    host = document.currentScript.getAttribute('data-host')
+    proxy = document.currentScript.getAttribute('data-proxy')
+    token = document.currentScript.getAttribute('data-token')
+    domain = document.currentScript.getAttribute('data-domain')
+    DATASOURCE =
+      document.currentScript.getAttribute('data-datasource') || DATASOURCE
+    STORAGE_METHOD =
+      document.currentScript.getAttribute('data-storage') || STORAGE_METHOD
+    for (const attr of document.currentScript.attributes) {
+      if (attr.name.startsWith('tb_')) {
+        globalAttributes[attr.name.slice(3)] = attr.value
+      }
     }
+  }
 
-    /**
-     * Generate uuid to identify the session. Random, not data-derived
-     */
-    function _uuidv4() {
-        return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
-            (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
-        );
+  /**
+   * Generate uuid to identify the session. Random, not data-derived
+   */
+  function _uuidv4() {
+    return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, c =>
+      (
+        c ^
+        (crypto.getRandomValues(new Uint8Array(1))[0] & (15 >> (c / 4)))
+      ).toString(16)
+    )
+  }
+
+  function _getSessionIdFromCookie() {
+    let cookie = {}
+    document.cookie.split(';').forEach(function (el) {
+      let [key, value] = el.split('=')
+      cookie[key.trim()] = value
+    })
+    return cookie[STORAGE_KEY]
+  }
+
+  function _getSessionId() {
+    if (
+      [storageMethods.localStorage, storageMethods.sessionStorage].includes(
+        STORAGE_METHOD
+      )
+    ) {
+      const storage =
+        STORAGE_METHOD === storageMethods.localStorage
+          ? localStorage
+          : sessionStorage
+      const serializedItem = storage.getItem(STORAGE_KEY)
+
+      if (!serializedItem) {
+        return null
+      }
+
+      try {
+        const item = JSON.parse(serializedItem)
+      } catch (error) {
+        return null
+      }
+
+      const now = new Date()
+
+      if (now.getTime() > item.expiry) {
```

Comment:
When testing local storage, I get an error on this line:

```
flock.js?v=3:78 Uncaught (in promise) ReferenceError: item is not defined
    at _getSessionId (flock.js?v=3:78:33)
    at _setSessionId (flock.js?v=3:110:27)
    at _sendEvent (flock.js?v=3:182:9)
    at flock.js?v=3:241:13
```

This is because `item` is defined inside the try block

Reply:
Can you check now?